src/arch/x86_64/Lower.zig

```diff
@@ -0,0 +1,465 @@
+        },
+        .rm_sib, .rm_rip => .{
+            .{ .reg = inst.data.rx.r },
+            .{ .mem = lower.mem(inst.ops, inst.data.rx.payload) },
+        },
+        else => return lower.fail("TODO lower {s} {s}", .{ @tagName(inst.tag), @tagName(inst.ops) }),
+    };
+    try lower.emit(.none, switch (ops[0].bitSize()) {
+        32, 64 => switch (ops[1].bitSize()) {
+            32 => .movsxd,
+            else => .movsx,
+        },
+        else => .movsx,
+    }, &ops);
+}
+
+fn mirCmovcc(lower: *Lower, inst: Mir.Inst) Error!void {
+    switch (inst.ops) {
+        .rr_cc => try lower.emit(.none, mnem_cc(.cmov, inst.data.rr_cc.cc), &.{
+            .{ .reg = inst.data.rr_cc.r1 },
+            .{ .reg = inst.data.rr_cc.r2 },
+        }),
+        .rm_sib_cc, .rm_rip_cc => try lower.emit(.none, mnem_cc(.cmov, inst.data.rx_cc.cc), &.{
+            .{ .reg = inst.data.rx_cc.r },
+            .{ .mem = lower.mem(inst.ops, inst.data.rx_cc.payload) },
+        }),
+        else => return lower.fail("TODO lower {s} {s}", .{ @tagName(inst.tag), @tagName(inst.ops) }),
+    }
+}
+
+fn mirSetcc(lower: *Lower, inst: Mir.Inst) Error!void {
+    switch (inst.ops) {
+        .r_cc => try lower.emit(.none, mnem_cc(.set, inst.data.r_cc.cc), &.{
+            .{ .reg = inst.data.r_cc.r },
+        }),
+        .m_sib_cc, .m_rip_cc => try lower.emit(.none, mnem_cc(.set, inst.data.x_cc.cc), &.{
+            .{ .mem = lower.mem(inst.ops, inst.data.x_cc.payload) },
+        }),
+        else => return lower.fail("TODO lower {s} {s}", .{ @tagName(inst.tag), @tagName(inst.ops) }),
+    }
+}
+
+fn mirPushPopRegisterList(lower: *Lower, inst: Mir.Inst) Error!void {
+    const save_reg_list = lower.mir.extraData(Mir.SaveRegisterList, inst.data.payload).data;
+    const base = @intToEnum(Register, save_reg_list.base_reg);
+    var disp: i32 = -@intCast(i32, save_reg_list.stack_end);
+    const reg_list = Mir.RegisterList.fromInt(save_reg_list.register_list);
+    const callee_preserved_regs = abi.getCalleePreservedRegs(lower.target.*);
+    for (callee_preserved_regs) |callee_preserved_reg| {
+        if (!reg_list.isSet(callee_preserved_regs, callee_preserved_reg)) continue;
+        const reg_op = Operand{ .reg = callee_preserved_reg };
+        const mem_op = Operand{ .mem = Memory.sib(.qword, .{ .base = base, .disp = disp }) };
+        try lower.emit(.none, .mov, switch (inst.tag) {
+            .push_regs => &.{ mem_op, reg_op },
+            .pop_regs => &.{ reg_op, mem_op },
+            else => unreachable,
+        });
+        disp += 8;
+    }
+}
+
+fn mirLeaLinker(lower: *Lower, inst: Mir.Inst) Error!void {
+    const metadata = lower.mir.extraData(Mir.LeaRegisterReloc, inst.data.payload).data;
+    const reg = @intToEnum(Register, metadata.reg);
+    try lower.emit(.none, .lea, &.{
+        .{ .reg = reg },
+        .{ .mem = Memory.rip(Memory.PtrSize.fromBitSize(reg.bitSize()), 0) },
+    });
+}
+
+const abi = @import("abi.zig");
```

Comment:
Again, please put them at the top of the file.

Reply:
This is something @andrewrk is experimenting with, it can always be moved later if we decide that it is bad.